doc/compiler/aot/Diagnostics.md

```diff
@@ -0,0 +1,396 @@
+                        tempDataStart: 00007F56828012C4, _aotMethodHeaderEntry: 00007F56828012CC, header offset: 8, binaryReloRecords: 00007F568280132C
+```
+This section contains much of the same information as before, but some
+additional information about the state of the relocation infrastructure
+prior to relocation.
+
+#### Validation / Relocation Trace
+```
+TR_InlinedVirtualMethodWithNopGuard 00007F5682801334
+        size 32 type 34 flags 0 reloFlags 3
+        inlined site index 0000000000000000
+        constant pool 00007F46DD4FFA18
+        cpIndex 000000000000001C
+        romClassOffsetInSharedCache 42f70 jdk/internal/misc/Unsafe
+        destinationAddress 00007F4714DE7100
+        validateSameClasses: caller method 00000000000AAAE0
+        inlinedSiteValid: caller method java/util/concurrent/ConcurrentHashMap.tabAt([Ljava/util/concurrent/ConcurrentHashMap$Node;I)Ljava/util/concurrent/ConcurrentHashMap$Node;
+        inlinedSiteValid: cp 00000000000A9720
+        getMethodFromCP: found virtual method 0000000000050E58
+        inlinedSiteValid: compileRomClass 00007F567147A7B8 currentRomClass 00007F567147A7B8
+        inlinedSiteValid: inlined method jdk/internal/misc/Unsafe.getObjectAcquire(Ljava/lang/Object;J)Ljava/lang/Object;
+        fixInlinedSiteInfo: [00007F56A26DE410] set to 0000000000050E58, virtual guard address 00007F56B8BC62D0
+        preparePrivateData: ramMethod 0000000000050E58 inlinedSiteIsValid 1
+        preparePrivateData: guard backup destination 00007F56A8BB69C0
+        reloLocation: from 00007F5682801364 at 00007F56A8BB68FE (offset 7e)
+                applyRelocation: activating inlined method
+```
+This section will vary depending on the relocation relocation. It is
+worth noting that the data printed out at the start of this section
+matches what was printed in the compile run. Also worth noting is
+the relocation line:
+```
+        reloLocation: from 00007F5682801364 at 00007F56A8BB68FE (offset 7e)
+```
+Again, the offset here is the `Offset (M)`. If a relocation record
+applies to multiple offsets, then there will be multiple lines printed.
+`00007F5682801364` is the instance of the `TR_RelocationRecord` type
+(or subtype); `00007F56A8BB68FE` is the address in the code the
+relocation is applied to. Not all relocations involve updating the
+code; some will use the relocated address to do other tasks such as
+registering Runtime Assumptions.
+
+# Common reasons for AOT bugs
+
+### Missing or incorrect relocation
+When adding a new feature, or enabling a disabled code path, it is
+important to think about AOT. If the compiler generates or emits
+something that is **not** position **independent**, then the compiled
+code may not work in another JVM instance. Depending on the position
+dependent code, the error may manifest itself in a way that is not
+trivial to understand. An example of this might be if the compiler
```

Comment:
I wonder just in case if you know an old issue that can demonstrate the example, maybe a link to the issue for others to take a look. Same comment for the next two common issues on examples.

Reply:
Update in force push.